### DISCUSSION TITLE
What is the Easiest Way to Add Driver Support for Analog/Discrete Input and Custom Routine?

### DISCUSSION THREAD

**Discussion by bear-fighter** [grblHAL/core#257] [Q&A]
I would like to add either an analog or discrete (PWM) input to the RP2040 so I can a write closed-loop spindle speed control routine.  The inner grblHAL workings of periodic ISR calls are still unclear after studying the source code.  What is the right level (i.e. source files) to add a routine?  Should I be setting up PWM timers at the driver level and registering my ISR that way or does grblHAL provide an interface for adding custom functions to the main operating loop (MOL)?

**Comment by terjeio** [comment]
> What is the right level (i.e. source files) to add a routine?

If you have a board map file then supporting code for that - like e.g. the [citoh_cx6000](https://github.com/grblHAL/RP2040/blob/13c7942173b28b56faff166450d101107c8b191d/citoh_cx6000_map.h#L35) has [here](https://github.com/grblHAL/RP2040/blob/master/citoh_cx6000.c). If not then a [user plugin](https://github.com/grblHAL/Templates/tree/master/my_plugin)?

> Should I be setting up PWM timers at the driver level and registering my ISR that way or does grblHAL provide an interface for adding custom functions to the main operating loop (MOL)?

You decide - both are possible. [Here is an example](https://github.com/grblHAL/Templates/blob/master/my_plugin/MCU%20load/my_plugin.c) for how to insert a function into the main loop. The yet to be completed [documentation](http://svn.io-engineering.com/grblHAL/html/hal_8h.html) may provide some details.
